### PR TITLE
Custom Feedback Paths: Sanitization and Validation

### PR DESCRIPTION
**Pull request by JosephineKwa**
1. Sanitized giver and recipient for use as spreadsheet data
2. Added backend validation for feedback paths

Valid feedback paths satisfy the following conditions:
- The givers can only be of a single giver type, i.e either a student, instructor or team
- The recipients can only be of a single recipient type, i.e either a student, instructor, team or class
- The giver and recipient both exist for the course that the feedback session is for
- No duplicate feedback paths
- Feedback paths are not partially filled, i.e. all rows should contain both a giver and a recipient. (Empty or partially filled paths are currently removed by the code)
- For team contribution questions, if a student is a giver, all the student's team members must also be givers. Each giver should have all the members in his/her team as a recipient. Both the giver and recipient have to be a student.

### CODE REVIEW

File: src/main/java/teammates/ui/controller/InstructorFeedbackQuestionEditAction.java [hunk 1]
```diff
@@ -141,6 +157,126 @@ public static String validateQuestionGiverRecipientVisibility(FeedbackQuestionAt
         
         return errorMsg;
     }
+    
+    public static String validateQuestionFeedbackPathsParticipants(
```
Comment (JosephineKwa) [inline src/main/java/teammates/ui/controller/InstructorFeedbackQuestionEditAction.java]: Will refactor to apply SLAP

**Comment by JosephineKwa** [comment]
@wkurniawan07 hope you don't mind!

File: src/main/java/teammates/common/util/FieldValidator.java [hunk 2]
```diff
@@ -658,9 +671,54 @@ private String getInvalidtyInfoForFirstTimeIsBeforeSecondTime(Date earlierTime,
...
         
+        if (giverType == FeedbackParticipantType.CUSTOM && recipientType != FeedbackParticipantType.CUSTOM
+                || recipientType == FeedbackParticipantType.CUSTOM && giverType != FeedbackParticipantType.CUSTOM) {
+            errors.add(PARTICIPANT_TYPE_CUSTOM_ERROR_MESSAGE);
+        }
+        
         return errors;
     }
     
+    public List<String> getValidityInfoForFeedbackPaths(List<FeedbackPathAttributes> feedbackPaths) {
```
Comment (wkurniawan07) [inline src/main/java/teammates/common/util/FieldValidator.java]: Supposed to be `getInvalidityInfo*`?

File: src/main/java/teammates/ui/controller/InstructorFeedbackQuestionAddAction.java [hunk 3]
```diff
@@ -78,6 +89,13 @@ protected ActionResult execute() {
     private String validateQuestionGiverRecipientVisibility(FeedbackQuestionAttributes feedbackQuestion) {
         return InstructorFeedbackQuestionEditAction.validateQuestionGiverRecipientVisibility(feedbackQuestion);
     }
+    
+    private String validateQuestionFeedbackPathsParticipants(
+            FeedbackQuestionAttributes question, List<StudentAttributes> students,
+            List<InstructorAttributes> instructors) {
+        return InstructorFeedbackQuestionEditAction.validateQuestionFeedbackPathsParticipants(
+                question, students, instructors);
```
Comment (wkurniawan07) [inline src/main/java/teammates/ui/controller/InstructorFeedbackQuestionAddAction.java]: This dependency is quite weird. Any way to extract the method somewhere else?

**Comment by damithc** [comment]
What's the status of this PR?

**Comment by wkurniawan07** [comment]
It looks like we don't have enough resources to take this big feature any further. I'm really sorry about this. Alternatively, if anyone is interested to continue the work done here, feel free to reopen/request for reopen.

**Pull request closed without merging**